Hotel 4.1

dudarenko-dmitry opened the pull request.

dudarenko-dmitry committed:
Hotel_4.1:
Updated README.md

dudarenko-dmitry committed:
Hotel_4.1:
Updated README.md

dudarenko-dmitry committed:
Merge remote-tracking branch 'origin/Hotel_4.1' into Hotel_4.1

dudarenko-dmitry committed:
Hotel_4.1:
Renamed Client to Guest, changed all Guest's layers.
Created objects Rooms, Guests and Reservations for further test of methods.

dudarenko-dmitry committed:
Hotel_4.1:
Removed IdHotelServices from HotelService.java to RoomReservation.java, GuideTour.java and GuideTour.
Made the appropriate changes in Entities, Repository and Service layers.

dudarenko-dmitry committed:
Hotel_4.1:
Removed IdHotelServices from HotelService.java to RoomReservation.java, GuideTour.java and GuideTour.
Made the appropriate changes in Entities, Repository and Service layers.

dudarenko-dmitry committed:
Hotel_4.1:
Created blueprints of interfaces with additional operations for all Entities on all layers.
Created empty interfaces-markers which extends CRUD-ALL operations and additional operations for all Entities on all layers.

dudarenko-dmitry committed:
Hotel_4.1:
Created 3 Comparators for Room and realised 3 corresponding methods:
- readAllRoomsSortByPrice,
- readAllRoomsSortByCapacity,
- readAllRoomsSortByLevel;

dudarenko-dmitry committed:
Hotel_4.1:
Created new entity FreeRoom and DataStorageFreeRoom.
Edited corresponding Controller, Service, Repository.
Edited enum RoomStatus.

dudarenko-dmitry committed:
Hotel_4.1:
Merged work with RoomReservation and FreeRoom in ServiceRoomReservation and ControllerRoomReservation.
Edited corresponding Controller, Service, Repository.
Edited enum RoomStatus.
Created 3 Comparators for FreeRoom and realised 3 corresponding methods:
- readAllFreeRoomsSortByPrice,
- readAllFreeRoomsSortByCapacity,
- readAllFreeRoomsSortByLevel;

dudarenko-dmitry committed:
Hotel_4.1:
Created 2 Comparators for RoomReservation and realised 3 corresponding methods:
- readAllRoomReservationsSortByGuestName,
- readAllRoomReservationsSortByGuestCheckOut;

dudarenko-dmitry committed:
Hotel_4.1:
Removed attribute StartDate from HotelService to RoomReservation, Restaurant, GuideTour.
Created HotelConstant with Start of Period, End of Period, Check-in Time, Check-out Time.
Changed in RoomReservation type of attributes from LocalDate to LocalDateTime.
Added method: countFreeRoomsOnTime(LocalDateTime).

dudarenko-dmitry committed:
Hotel_4.1:
Added method: countNumberOfGuestsTotal().

dudarenko-dmitry committed:
Hotel_4.1:
Added method: countNumberOfGuestsOnDate().

dudarenko-dmitry committed:
Hotel_4.1:
Added method: readAllRoomsFreeAtTime(LocalDateTime time).

dudarenko-dmitry committed:
Hotel_4.1:
Added method: countGuestPaymentForRoom(int idGuest).

dudarenko-dmitry committed:
Hotel_4.1:
Added method: read3LastGuestAndDatesForRoom(int idRoom).

mrseneca commented on src/pl/senla/hotel/controller/ControllerRoomReservation.java:
```diff
@@ -0,0 +1,7 @@
+
+import pl.senla.hotel.entity.RoomReservation;
+
+public interface ControllerRoomReservation extends ControllerCRUDALL<RoomReservation>,
```
Let's not create these intermediate interfaces, let's use multiple implementations in classes. Because now you have 4 classes in this package, there are 12 interfaces. It is very problematic to read and understand the code. I had to download the repository to the PC

dudarenko-dmitry replied:
which is optimal quantity of interfaces for 4 classes?

mrseneca replied:
it's hard to tell, to be honest :D
Look, your code should always be simple, easy to read and cover the task. Now you are trying to cover all the following developments of your application by creating abstract interfaces and inheriting them. Right now it is unnecessary and redundant, the code is too cumbersome and difficult to understand. Right now, you can make separate interfaces for each class, and it will be ok even if they have duplicate methods, but with different return types. But, I'm not trying to wean you from optimization, what you tried to do well, but it turned out superfluous. Let's do this, for now, make one common interface with СRUD operations and 4 additional ones that will extend it with the methods necessary for classes. And only then, when the project grows, you can either expand or narrow the abstraction

mrseneca replied:
Similar approach to repositories also

dudarenko-dmitry replied:
I'try to explain for which reason I created intermediate interfaces.

For example Repository for entity Guest: 
before: **RepositoryGuestImpl implements RepositoryGuest**.
after: **RepositoryGuestImpl implements Repository<Guest>, RepositoryAdditional**.
We exclude intermediate interface RepositoryGuest which extended earlier 2 above mentioned interfaces.
So we will reduce total quantity of interfaces in package from 12 at least to 5-7.

But in service layer in this case to save incupsulation I must create 2 instances which work with methods from RepositoryGuestImpl.
before:
    private final RepositoryGuest guestRepository = new RepositoryGuestCollection();
after:
    private final RepositoryCRUDALL<Guest> guestRepositoryCRUDALL = new RepositoryGuestCollection();
    private final RepositoryGuestAdditional guestRepositoryAdditional = new RepositoryGuestCollection();
It's not a problem. I can refactor this in 1 hour.

I can go in such ways:
1. I'll delete intermadiate interfaces and add additional entities to upper layers (as it is shown above) if I understand your proposal correctly;
2. I'll ave present quantity of interfaces but divide them to different subpackages (repository.guest, repository.reservation, etc.). It will make structure easier readable.
3. Perhaps you could offer another more efficient way on base of your experience.

dudarenko-dmitry replied:
ready

mrseneca commented on src/pl/senla/hotel/repository/RepositoryFreeRoom.java:
```diff
@@ -0,0 +1,6 @@
+
+import pl.senla.hotel.entity.FreeRoom;
+
+public interface RepositoryFreeRoom extends RepositoryCRUDALL<FreeRoom>, RepositoryFreeRoomAdditional {
```
Same as with Controllers

dudarenko-dmitry replied:
ready

mrseneca commented on src/pl/senla/hotel/entity/FreeRoomComparatorByPrice.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.Comparator;
+
+public class FreeRoomComparatorByPrice implements Comparator<FreeRoom> {
```
Let's move comparators to separate package - .utils or .comparators or something like that. Because technically  they are not entyties

dudarenko-dmitry replied:
ready

mrseneca reviewed (commented):
Good, but hard to read. Let's fix this for now, and I'll check further

dudarenko-dmitry committed:
Hotel_4.1.
Made optimisation in application structure:
1. moved methods from additional interfaces to "base"-interfaces;
2. deleted additional interfaces;
3. moved all Comparators to brand-new package .comparators.

dudarenko-dmitry commented:
Please, do not approve commit 1-2 days.
I want to finish 9,10,11 from TASK_4.1 and refactor creating of new RoomReservations (need additional checking if room is free at current time) before branch's merging.

mrseneca commented on src/pl/senla/hotel/repository/RepositoryOrderCollection.java:
```diff
@@ -44,4 +46,14 @@ public boolean delete(int id) {
+
+    @Override
+    public List<HotelService> readAllServicesSortByPrice(int idGuest) {
+        return readAll().stream().filter(hs -> hs.getClient().getIdGuest() == idGuest).flatMap(o -> o.getServices().stream()).sorted(new HotelServicesComparatorByPrice()).toList();
```
Can't find this comparator anywhere. Can you check please, if it was pushed to repo?

dudarenko-dmitry replied:
This task is in work. So I haven't still pushed it.

Perhaps there will be some questions because HotelService is an abstract class and it has only 1 attribute inside (Guest guest).
Price and Dates/Times I moved to its child classe. May be it's mistake... And it is better to return these attributes back to parent class.

I wrote such Comparator:
public class HotelServicesComparatorByPrice implements Comparator<HotelService> {
    @Override
    public int compare(HotelService o1, HotelService o2) {
        switch(o1.getTypeOfService()){
            case "RoomReservation":
                return compareRoomReservation((RoomReservation) o1, (RoomReservation) o2);
            case "Restaurant":
                return compareRestaurant((Restaurant) o1, (Restaurant) o2);
            case "GuideTour":
                return compareGuideTour((GuideTour) o1, (GuideTour) o2);
            default:
                System.out.println("ERROR IN SERVICE TYPE.");
        }
        return 0;
    }

    private int compareRoomReservation(RoomReservation o1, RoomReservation o2){
        return o1.getRoom().getRoomPrice() - o2.getRoom().getRoomPrice();
    }

    private int compareRestaurant(Restaurant o1, Restaurant o2){
        return o1.getPrice() - o2.getPrice();
    }

    private int compareGuideTour(GuideTour o1, GuideTour o2){
        return o1.getPrice() - o2.getPrice();
    }
}

It looks too "heavy" but it works (tested with RoomReservations only).

mrseneca commented on src/pl/senla/hotel/controller/ConsoleCollection.java:
```diff
@@ -3,92 +3,210 @@
-        Room room2 = new Room(2,13,130,2,RoomLevel.LUX.getLevel(), RoomStatus.FREE.getStatus());
+        System.out.println("\n----- Room and FreeRoom -----");
+        System.out.println("==============================");
+        System.out.println("CREATE and READ ROOMS");
```
Let's reduce amount of basic entities (guest, room) to 4-5, it will also reduce dependent entities. 
It's to much right now, the are similar but create a wall of code and logs

mrseneca replied:
Or if you want have all current data, try to optimize code (streams, lambdas, switches) to make it less duplicated

dudarenko-dmitry replied:
I used only these basic Entities:
1. Room
2. Guest
3. RoomReservation (it is connected with Room and Guest)
4. FreeRoom (it's the opposite of a RoomReservation and will be used for checking if the creating of new RoomReservation is available). My opinion: It's impossible to delete this entity.
5. HotelService (it's parent/abstract class for all types of Hotel Services and is used for creating of List of all Guest's Services in Order).
6. Order (it serves to connection of all different types of Orders which Guest reserved).
------------------------
7 and 8. Restaurant and GuideTour - these HotelServices were created as examples of other HotelServices but I don't use them.

_**"Or if you want have all current data, try to optimize code (streams, lambdas, switches) to make it less duplicated."**_
ok. I use them...

Should I use Builder for creating new instances?

mrseneca replied:
No no, entities are ok, I meant amount of created entities, you have more then 10 room and more then 10 guests. Let's reduce it to 4-5. It's to much to read it in code and logs

mrseneca replied:
You can use builders if you understand them. Right now it's not necessary, but totally acceptable

mrseneca commented on src/pl/senla/hotel/controller/ConsoleCollection.java:
```diff
@@ -3,92 +3,210 @@
-        System.out.println(CONSOLE_READ_ALL_CLIENTS + clientController.readAll());
+        System.out.println("\n============================");
+        System.out.println("UPDATE GUEST");
+        guest0.setPhoneNumber(378340012); //update Client
```
It will work here, because link;s will be same and you working in one thread. But for future remember - you must receive entity you want to change from entity holder. 
No need to fix it, just reminder for future

dudarenko-dmitry committed:
Hotel_4.1.
Added method readAllServicesSortByPrice(int idGuest).

dudarenko-dmitry committed:
Hotel_4.1.
Added method readAllServicesSortByDate(int idGuest) and HotelServicesComparatorByDate.
.

dudarenko-dmitry committed:
Hotel_4.1.
Added new classes Table, Transport, abstract class HotelFacility from which Room, Table and Transport are extended.
Created Interfaces and their implementation on all layers and 2 Comparators for HotelFacility.
Added methods:
readPriceListForServicesSortByCategory();
readPriceListForServicesSortByPrice().
Created in package .entity new sub-packages to make structure more readable.

dudarenko-dmitry committed:
Hotel_4.1.
Made refactoring of method create(RoomReservation reservation): added checking if current Room is available at selected time.

dudarenko-dmitry committed:
Hotel_4.1.
Made refactoring of method delete(RoomReservation reservation): added of new FreeRoom at time of deleted RoomReservation.

dudarenko-dmitry committed:
Hotel_4.1.
Made refactoring of method update(RoomReservation reservation):
1. added/deleted of new FreeRoom at time of deleted RoomReservation.
2. added logic for automatic change of Room's Reservations' cost.

dudarenko-dmitry committed:
Hotel_4.1.
Refactored creation of RoomReservations: added method for ID auto-generation.

dudarenko-dmitry committed:
Hotel_4.1.
Refactored creation of FreeRoom: added method for ID auto-generation.
Refactored method for ID auto-generation for RoomReservation.

dudarenko-dmitry committed:
Hotel_4.1.
Refactored creation of Room: added method for ID auto-generation.

dudarenko-dmitry committed:
Hotel_4.1.
Refactored creation of Guest: added method for ID auto-generation.

dudarenko-dmitry committed:
Hotel_4.1.
Refactored creation of Order: added method for ID auto-generation.
Small refactoring of other classes and methods.

dudarenko-dmitry committed:
Deleted package with Lecture 3

mrseneca reviewed (approved):
It will be it. But for future read about git commit squash feature, it may be crucial for some teams and projects

dudarenko-dmitry committed:
TASK_4.2.
Create hotel.jar.
Create hotel.bat.
